### PR TITLE
support equal binary predicate on zone map filter

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Currently, we can't apply equal binary predicate on zone map filter because it is not monotonic, you can see more details in the [previous PR](https://github.com/StarRocks/starrocks/pull/2520). In fact, `=` is equivalent to `>= and <=`, both `>=` and `<=` are monotonic, so we can apply `=` on zone map filter by rewriting it to `>= and <=`. After this change, some predicates like `year(col) = 2000` will have a chance to use zone map filter.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 4 / 4 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|2|2|100.00%|

File: be/src/storage/vectorized/column_expr_predicate.cpp [hunk 1]
```diff
@@ -176,6 +177,70 @@ std::string ColumnExprPredicate::debug_string() const {
...
+                                                               std::vector<const ColumnExprPredicate*>* output) {
+    DCHECK(pool != nullptr);
+    DCHECK(output != nullptr);
+    ExprContext* pred_from_planner = _expr_ctxs[0];
+    Expr* root = pred_from_planner->root();
+
+    std::vector<Expr*> exprs_after_rewrite;
+
+    if (root->op() == TExprOpcode::EQ) {
+        DCHECK(root->get_num_children() == 2);
```
Comment (mofeiatwork) [inline be/src/storage/vectorized/column_expr_predicate.cpp:190]: Maybe need an explicit check?

File: be/src/storage/vectorized/column_expr_predicate.cpp [hunk 2]
```diff
@@ -176,6 +177,70 @@ std::string ColumnExprPredicate::debug_string() const {
...
+    // build new ColumnExprPredicates
+    for (auto& expr : exprs_after_rewrite) {
+        ExprContext* ctx = new ExprContext(expr);
+        RETURN_IF_ERROR(ctx->prepare(_state));
+        RETURN_IF_ERROR(ctx->open(_state));
+        ColumnExprPredicate* new_pred = new ColumnExprPredicate(_type_info, _column_id, _state, ctx, _slot_desc);
+        // copy other cast exprs
+        for (size_t i = 1; i < _expr_ctxs.size(); i++) {
+            Expr* tmp_expr = Expr::copy(pool, _expr_ctxs[i]->root());
+            ExprContext* tmp_ctx = new ExprContext(tmp_expr);
```
Comment (mofeiatwork) [inline be/src/storage/vectorized/column_expr_predicate.cpp:230]: Would it be leaked if return_if_error?

File: be/src/storage/vectorized/column_predicate_rewriter.cpp [hunk 3]
```diff
@@ -391,4 +391,41 @@ void ConjunctivePredicatesRewriter::rewrite_predicate(ObjectPool* pool) {
...
+            dst->emplace_back(pred);
+        }
+    }
+    return Status::OK();
+}
+
+Status ZonemapPredicatesRewriter::_rewrite_column_expr_predicates(ObjectPool* pool, const ColumnPredicate* pred,
+                                                                  std::vector<const ColumnExprPredicate*>* new_preds) {
+    DCHECK(new_preds != nullptr);
+    ColumnExprPredicate* column_expr_pred =
```
Comment (mofeiatwork) [inline be/src/storage/vectorized/column_predicate_rewriter.cpp:426]: Can we eliminate this wired cast?

File: be/src/storage/vectorized/column_expr_predicate.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/storage/vectorized/column_expr_predicate.cpp]: > Would it be leaked if return_if_error?

fixed

File: be/src/storage/vectorized/column_predicate_rewriter.cpp (hunk 3 above)
Comment (silverbullet233) [reply be/src/storage/vectorized/column_predicate_rewriter.cpp]: > Can we eliminate this wired cast?

fixed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 4 / 4 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|2|2|100.00%|

File: be/src/storage/vectorized/column_expr_predicate.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/storage/vectorized/column_expr_predicate.cpp:190]: > Maybe need an explicit check?

what do you mean by `explicit check`? if `root->get_num_children() != 2` , we just return?

Comment (mofeiatwork) [reply be/src/storage/vectorized/column_expr_predicate.cpp:190]: yes

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 4]
```diff
@@ -460,6 +461,10 @@
     @VariableMgr.VarAttr(name = ENABLE_COLUMN_EXPR_PREDICATE)
     private boolean enableColumnExprPredicate = false;
 
+    @VariableMgr.VarAttr(name = ENABLE_REWRITE_ZONE_MAP_PREDICATE)
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:464]: invisible

File: be/src/storage/vectorized/column_predicate_rewriter.h [hunk 5]
```diff
@@ -77,4 +79,18 @@ class ConjunctivePredicatesRewriter {
     std::vector<uint8_t>* _disable_dict_rewrite;
 };
 
+class ZonemapPredicatesRewriter {
```
Comment (kangkaisen) [inline be/src/storage/vectorized/column_predicate_rewriter.h:82]: comment why and when we need rewrite.

File: be/src/storage/vectorized/column_predicate_rewriter.h [hunk 6]
```diff
@@ -77,4 +79,18 @@ class ConjunctivePredicatesRewriter {
...
+public:
+    using PredicateList = std::vector<const ColumnPredicate*>;
+
+    static Status rewrite(ObjectPool* pool, const std::unordered_map<ColumnId, PredicateList>& src,
```
Comment (kangkaisen) [inline be/src/storage/vectorized/column_predicate_rewriter.h:86]: what's the difference of two `rewrite` method. would better give a more concrete method name

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 4 above)
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:464]: I think would better directly remove this variable if the default value is true, there are too many codes for this flag.

**Review by kangkaisen — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Pull request merged**